Open Write

seanmcc-msft opened the pull request:
Continued from https://github.com/Azure/azure-sdk-for-net/pull/12779 and https://github.com/Azure/azure-sdk-for-net/pull/12662

seanmcc-msft committed:
Open Write

seanmcc-msft commented:
@PaulVrugt, we are continuing Open Write here.

PaulVrugt commented:
Cool, thanks for the heads up!

seanmcc-msft committed:
Added unit test for PooledMemoryStream

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/BlockBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,37 @@
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
+        /// max is 100 MB.
```
reference the block blob client property for this

seanmcc-msft replied:
Same as above.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/PageBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        public bool Overwrite { get; set; }
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
+        /// max is 4 MB.  Must be a increment of 512.
```
More referencing exposed properties that describe these

seanmcc-msft replied:
Same as above.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileOpenWriteOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        public bool Overwrite { get; set; }
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
+        /// max is 4 MB.
```
Reference exposed properties.

seanmcc-msft replied:
Same as above.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileOpenWriteOptions.cs:
```diff
@@ -0,0 +1,43 @@
+
+        /// <summary>
+        /// Required if Overwrite is set to true.
+        /// Specifies the size of the new Page Blob.
```
File, not page blob. Also the comment about attempting to group this with overwrite.

seanmcc-msft replied:
Fixed page blob -> file.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileOpenWriteOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        /// Required if Overwrite is set to true.
+        /// Specifies the size of the new Page Blob.
+        /// </summary>
+        public long MaxSize { get; set; }
```
Page blob and datalake just used `Size`. Is this more common terminology in file shares? Are files actually not always the size they're given?

seanmcc-msft replied:
SMB Files have a max size, which is specified when you [create them](https://docs.microsoft.com/en-us/rest/api/storageservices/create-file#request-headers).

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4385,5 +4385,144 @@ private void SetNameFieldsIfNull()
+        /// a failure occurs.
+        /// </remarks>
+        private async Task<Stream> OpenWriteInternal(
+            bool async,
+            long position,
+            ShareFileOpenWriteOptions options,
+            CancellationToken cancellationToken)
```
position, options, async, cancellationToken

seanmcc-msft replied:
Fixed in next iteration.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4385,5 +4385,144 @@ private void SetNameFieldsIfNull()
+
+                if (options != null && options.Overwrite)
+                {
+                    if (async)
+                    {
+                        await CreateAsync(
+                            maxSize: options.MaxSize,
+                            conditions: options.Conditions,
+                            cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Create(
+                            maxSize: options.MaxSize,
+                            conditions: options.Conditions,
+                            cancellationToken: cancellationToken);
+                    }
```
`CreateInternal`

seanmcc-msft replied:
Fixed in next iteration.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4385,5 +4385,144 @@ private void SetNameFieldsIfNull()
+            {
+                scope.Start();
+
+                if (options != null && options.Overwrite)
```
`options?.Overwrite`

seanmcc-msft replied:
Fixed in next iteration.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4385,5 +4385,144 @@ private void SetNameFieldsIfNull()
+                }
+                return new ShareFileWriteStream(
+                    fileClient: this,
+                    bufferSize: options?.BufferSize ?? Constants.MB,
```
Use a defined constant for buffer size.

seanmcc-msft replied:
Fixed in next iteration.

jaschrep-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileWriteStream.cs:
```diff
@@ -0,0 +1,83 @@
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 100 * Constants.MB)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must <= 4 MB");
+            }
```
Use max file write constant.
The check size and the message don't match.

seanmcc-msft replied:
Fixed in next iteration.

jaschrep-msft reviewed (commented):
Mostly nitpicking, but there are some potential bugs to be addressed. Also some questions on tiny design elements (nothing major).
Will review tests when they are completed.

seanmcc-msft committed:
Merge branch 'master' into feature/storage/openWrite

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -890,6 +898,14 @@ public partial class BlobUploadOptions
+    {
+        public BlockBlobOpenWriteOptions() { }
+        public long BufferSize { get { throw null; } set { } }
+        public Azure.Storage.Blobs.Models.BlobRequestConditions Conditions { get { throw null; } set { } }
```
No, the other `BlockBlobClient` APIs use `BlobRequestConditions`.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1341,6 +1341,147 @@ public new AppendBlobClient WithVersion(string versionId)
+        /// a failure occurs.
+        /// </remarks>
+        private async Task<Stream> OpenWriteInternal(
+            bool async,
+            AppendBlobOpenWriteOptions options,
+            CancellationToken cancellationToken)
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1341,6 +1341,147 @@ public new AppendBlobClient WithVersion(string versionId)
+
+                long position = 0;
+
+                if (options != null && options.Overwrite)
```
`options?.Overwite` results in a nullable boolean, which the if statement does not like.  I changed it to `options?.Overwrite == true`.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1341,6 +1341,147 @@ public new AppendBlobClient WithVersion(string versionId)
+
+                return new AppendBlobWriteStream(
+                    appendBlobClient: this,
+                    bufferSize: options?.BufferSize ?? Constants.MB,
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobWriteStream.cs:
```diff
@@ -0,0 +1,79 @@
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 4 * Constants.MB)
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -1949,6 +1949,151 @@ protected sealed override BlobBaseClient WithSnapshotCore(string snapshot)
+
+                long position = 0;
+
+                if (options != null && options.Overwrite)
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1341,6 +1341,147 @@ public new AppendBlobClient WithVersion(string versionId)
+
+                if (options != null && options.Overwrite)
+                {
+                    if (async)
+                    {
+                        await CreateAsync(
+                            conditions: options?.Conditions,
+                            cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Create(
+                            conditions: options?.Conditions,
+                            cancellationToken: cancellationToken);
+                    }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1341,6 +1341,147 @@ public new AppendBlobClient WithVersion(string versionId)
+                {
+                    BlobProperties blobProperties;
+
+                    if (async)
+                    {
+                        blobProperties = await GetPropertiesAsync().ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        blobProperties = GetProperties();
+                    }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -1949,6 +1949,151 @@ protected sealed override BlobBaseClient WithSnapshotCore(string snapshot)
+
+                if (options != null && options.Overwrite)
+                {
+                    if (async)
+                    {
+                        await UploadAsync(
+                            content: new MemoryStream(Array.Empty<byte>()),
+                            conditions: options.Conditions,
+                            progressHandler: options.ProgressHandler,
+                            cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Upload(
+                            content: new MemoryStream(Array.Empty<byte>()),
+                            conditions: options.Conditions,
+                            progressHandler: options.ProgressHandler,
+                            cancellationToken: cancellationToken);
+                    }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -1949,6 +1949,151 @@ protected sealed override BlobBaseClient WithSnapshotCore(string snapshot)
+                {
+                    BlobProperties blobProperties;
+
+                    if (async)
+                    {
+                        blobProperties = await GetPropertiesAsync().ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        blobProperties = GetProperties();
+                    }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -2801,6 +2801,146 @@ public new PageBlobClient WithVersion(string versionId)
+            {
+                scope.Start();
+
+                if (options != null && options.Overwrite)
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -2801,6 +2801,146 @@ public new PageBlobClient WithVersion(string versionId)
+
+                if (options != null && options.Overwrite)
+                {
+                    if (async)
+                    {
+                        await CreateAsync(
+                            size: options.Size,
+                            conditions: options.Conditions,
+                            cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Create(
+                            size: options.Size,
+                            conditions: options.Conditions,
+                            cancellationToken: cancellationToken);
+                    }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -2801,6 +2801,146 @@ public new PageBlobClient WithVersion(string versionId)
+
+                return new PageBlobWriteStream(
+                    pageBlobClient: this,
+                    bufferSize: options?.BufferSize ?? Constants.MB,
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,143 @@
+            else
+            {
+                // We need a multiple of 512 to flush.
+                if (_buffer.Length % 512 != 0)
+                {
+                    int bytesToWrite = (int)(512 - _buffer.Length % 512);
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,143 @@
+
+            if (bufferSize % 512 != 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be a multiple of 512");
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,143 @@
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 4 * Constants.MB)
```
Fixed in next iteration.

tg-msft replied:
Was it fixed?

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,111 @@
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 100 * Constants.MB)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must <= 100 MB");
+            }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,143 @@
+
+                while (remaining > 0)
+                {
+                    await WriteToBuffer(
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,143 @@
+
+                if (async)
+                {
+                    await _pageBlobClient.UploadPagesAsync(
+                        _buffer,
+                        _writeIndex,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _pageBlobClient.UploadPages(
+                        _buffer,
+                        _writeIndex,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs:
```diff
@@ -0,0 +1,198 @@
+
+namespace Azure.Storage.Shared
+{
+    internal abstract class WriteStream : Stream
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:
```diff
@@ -3591,6 +3591,147 @@ Response<Blobs.Models.BlobDownloadInfo> response
+
+                long position = 0;
+
+                if (options != null && options.Overwrite)
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:
```diff
@@ -3591,6 +3591,147 @@ Response<Blobs.Models.BlobDownloadInfo> response
+
+                if (options != null && options.Overwrite)
+                {
+                    if (async)
+                    {
+                        await CreateAsync(
+                            conditions: options.Conditions,
+                            cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        Create(
+                            conditions: options.Conditions,
+                            cancellationToken: cancellationToken);
+                    }
+                }
+                else
+                {
+                    PathProperties pathProperties;
+
+                    if (async)
+                    {
+                        pathProperties = await GetPropertiesAsync().ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        pathProperties = GetProperties();
+                    }
```
Fixed for `CreateInternal()`, but we have to call `GetProperties()` and `GetPropertiesAsync()` because they call into Blobs.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+            {
+                _buffer.Position = 0;
+
+                if (async)
+                {
+                    await _fileClient.AppendAsync(
+                        content: _buffer,
+                        offset: _writeIndex,
+                        leaseId: _conditions?.LeaseId,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _fileClient.Append(
+                        content: _buffer,
+                        offset: _writeIndex,
+                        leaseId: _conditions?.LeaseId,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            if (async)
+            {
+                await _fileClient.FlushAsync(
+                    position: _writeIndex,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                _fileClient.Flush(
+                    position: _writeIndex,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken);
+            }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > 100 * Constants.MB)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must <= 100 MB");
+            }
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/Models/DataLakeFileOpenWriteOptions.cs:
```diff
@@ -0,0 +1,44 @@
+        /// </summary>
+        public long BufferSize { get; set; }
+
+        /// <summary>
+        /// Required <see cref="Overwrite"/> is set to true.
+        /// Specifies the size of the new Page Blob.
+        /// </summary>
+        public long Size { get; set; }
```
No, this was a copy + paste error.  Page Blob and SMB Files have fixed sizes, Block Blobs, Append Blobs, and Data Lake Files do not.  Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/Models/DataLakeFileOpenWriteOptions.cs:
```diff
@@ -0,0 +1,44 @@
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
+        /// max is 100 MB.
```
I'm not sure we have a public property on any of the Data Lake clients that exposes this value.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/AppendBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,38 @@
+        public bool Overwrite { get; set; }
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
```
I don't thing we have this value public exposed anywhere.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/AppendBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,38 @@
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
+        /// max is 4 MB.
```
Do we publicly expose this value somewhere?

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/BlockBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,37 @@
+        public bool Overwrite { get; set; }
+
+        /// <summary>
+        /// The size of the buffer to use.  Default is 1 MB,
```
Same as above.

seanmcc-msft committed:
PR comments

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/PageBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        /// Whether the upload should overwrite any existing blobs.
+        /// The default value is false.
+        /// </summary>
+        public bool Overwrite { get; set; }
```
You Page Blob will be overwritten, and we will being writing the new content at offset 512.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,143 @@
+        public PageBlobWriteStream(
+            PageBlobClient pageBlobClient,
+            long bufferSize,
+            long position,
```
https://github.com/Azure/azure-sdk-for-net/pull/13344/files?file-filters%5B%5D=.cs#diff-e39f8ee2730163a6c9c2d682c1e9c025R127

jaschrep-msft replied:
Position, not buffer size. I don't see position getting passed to any validation method in this or the base constructor.

Regarding `ValidateBufferSize`, it's abstract. Why don't we call the method in the base constructor instead of repeatedly in each subclass?

seanmcc-msft replied:
Added `ValidatePosition()` to `PageBlobWriteStream` in the next iteration.

I don't call `ValidateBufferSize()` in the base constructor because:

`CA2214: Do not call overridable methods in constructors `.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,111 @@
+    {
+        private readonly BlockBlobClient _blockBlobClient;
+        private readonly BlobRequestConditions _conditions;
+        private readonly List<string> _blockIds;
```
Fixed in next iteration.

jaschrep-msft replied:
I don't feel super comfortable with that behavior. If we aren't overwriting and can't get a block list to append to, I'd rather throw instead of having the one case where we blow your data away even when specified not to.

seanmcc-msft replied:
Removing `BlockBlobOpenWriteOptions.Overwrite` in the next iteration, I intended to already.

jaschrep-msft replied:
Do other streams default to false with that option? I'd expect them to, and I feel like that's a gotcha if this one case always overwrites.

Under that assumption: unless we make all the other write stream options `bool? Overwrite` and throw if they left it null, I think we need to make some sort of attempt to do an append if we want to provide a consistent user experience. The only downside I can think of is the extra service call we have to wait on before uploading.

I prefer the attempt to append, throwing when the operation is impossible. I dislike mandating users to set a `bool?` and it gives users more possibilities that would otherwise be denied.

seanmcc-msft replied:
The other streams append by default.

@kasobol-msft, what do you think on this issue?  In track 1, `CloudBlockBlob.OpenWrite()` always overwrote the underlying blob.

kasobol-msft replied:
I'm looking at https://docs.microsoft.com/en-us/dotnet/api/system.io.file.openwrite?view=netcore-3.1 .
It seems that OpenWrite always overwrites if file exists. If user wants to append they're expected to use "File.AppendText". Therefore I think Overwrite should be default for all "OpenWrite" implementations.

Can we have a `bool` with reasonable default for `Overwrite`. I'm not sure we need 3-state boolean. https://thedailywtf.com/articles/tri-state-boolean

seanmcc-msft replied:
After syncing offline, we decided to remove the overwrite parameter entirely.  All `WriteStreams` except `BlockBlobWriteStream` will append, and `BlockBlobWriteStream` will overwrite.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,111 @@
+            }
+        }
+
+        private static string GenerateBlockId()
+        {
+            Guid guid = Guid.NewGuid();
+            byte[] bytes = Encoding.UTF8.GetBytes(guid.ToString());
+            return Convert.ToBase64String(bytes);
+        }
```
Do the block IDs really all have to be the same length?  This seems like a strange requirement.

seanmcc-msft replied:
Based on my previous comment about Put Blob not creating Blocks, I think we don't need to worry about this.

jaschrep-msft replied:
Block IDs do have to be the same length. The service mandates it due to significant performance optimizations it can achieve with that constraint.

Leaving this discussion open until agreement is reached on how to deal with the block list situation.

seanmcc-msft replied:
I've switched to using the same algorithm to generate Block IDs as partitioned upload.

seanmcc-msft committed:
PR comments

seanmcc-msft committed:
Added AppendBlobClient.OpenWrite() tests

seanmcc-msft committed:
Added PageBlobClient.OpenWrite() tests

seanmcc-msft committed:
Added ShareFileClient.OpenWrite() tests

seanmcc-msft committed:
Added DataLakeFileClient.OpenWrite() tests

seanmcc-msft committed:
Added BlockBlobClient.OpenWrite() tests

seanmcc-msft committed:
Ran Export-API.ps1

seanmcc-msft committed:
Recording tests

seanmcc-msft committed:
More test recording

seanmcc-msft committed:
should pass CI now

seanmcc-msft committed:
Removed BlockBlobOpenWriteOptions.Overwrite

seanmcc-msft committed:
Merge branch 'master' into feature/storage/openWrite

seanmcc-msft committed:
Hopefully fixed CI

seanmcc-msft committed:
Added PageBlobWriteStream.ValidatePosition()

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/PageBlobOpenWriteOptions.cs:
```diff
@@ -0,0 +1,43 @@
+    /// <summary>
+    /// Optional parameters for Page Blob Open Write.
+    /// </summary>
+    public class PageBlobOpenWriteOptions
```
I think there are enough differences that it's ok to have separate classes:
```
    public class AppendBlobOpenWriteOptions
    {
        public bool Overwrite { get; set; }
        public long? BufferSize { get; set; }
        public AppendBlobRequestConditions Conditions { get; set; }
        public IProgress<long> ProgressHandler { get; set; }
    }
```
```
    public class BlockBlobOpenWriteOptions
    {
        public long? BufferSize { get; set; }
        public BlobRequestConditions Conditions { get; set; }
        public IProgress<long> ProgressHandler { get; set; }
    }
```
```
    public class PageBlobOpenWriteOptions
    {
        public bool Overwrite { get; set; }
        public long? BufferSize { get; set; }
        public long Size { get; set; }
        public PageBlobRequestConditions Conditions { get; set; }
        public IProgress<long> ProgressHandler { get; set; }
    }
```

It looks like BufferSize and ProgressHandler are the only properties that match exactly.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,115 @@
+            }
+        }
+
+        private static string GenerateBlockId(long offset)
+        {
+            // TODO #8162 - Add in a random GUID so multiple simultaneous
+            // uploads won't stomp on each other and the first to commit wins.
+            // This will require some changes to our test framework's
+            // RecordedClientRequestIdPolicy.
+            byte[] id = new byte[48]; // 48 raw bytes => 64 byte string once Base64 encoded
+            BitConverter.GetBytes(offset).CopyTo(id, 0);
+            return Convert.ToBase64String(id);
+        }
+    }
```
One option would be to create `internal IBlockIdGenerator` and substitute with something predictable in tests.

seanmcc-msft replied:
This was copied from [BlockBlobClient.PartitionedUploader.GenerateBlockId()](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs#L2018)

kasobol-msft replied:
That'd be another reason to extract this logic from uploader to something in shared source so that both components can use the same.

jaschrep-msft replied:
This is already predictable. It was made that way to be producible in recorded tests. An object is probably the best way to share code, but I don't think we need to interface it or anything. Mocks are suitable enough for tests if we ever change this to a non-predictable system (we may need a new system if we support appends based on other discussion).

seanmcc-msft replied:
This was copy + pasted from the parallel uploader, including the comment.  I'm fine moving the static method to Common.

seanmcc-msft replied:
Moved to common in next iteration.

seanmcc-msft committed:
Hopefully fixed CI

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,115 @@
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            if (async)
+            {
+                await _blockBlobClient.CommitBlockListAsync(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                _blockBlobClient.CommitBlockList(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken);
+            }
+        }
```
await `CommitBlockListInternal`

seanmcc-msft replied:
Fixed.

jaschrep-msft reviewed (commented):
We should test `StorageWriteStream` on its own in Azure.Storage.Common.Tests, mocking out the abstract behavior. I'm particularly interested in testing edge cases on combining `Stream.WriteAsync` and `Stream.FlushAsync` calls, observing when/how they trigger `AppendInternal` calls that actually send data to the service.

seanmcc-msft committed:
PR comments

seanmcc-msft commented:
> We should test StorageWriteStream on its own in Azure.Storage.Common.Tests, mocking out the abstract behavior. I'm particularly interested in testing edge cases on combining Stream.WriteAsync and Stream.FlushAsync calls, observing when/how they trigger AppendInternal calls that actually send data to the service.

A lot of this logic is handled in the implementation streams, I'm not sure how valuable having tests just for StorageWriteStream will be.

seanmcc-msft committed:
Removed overwrite parameters

seanmcc-msft committed:
Added tests for StorageWriteStream

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/tests/StorageWriteStreamTests.cs:
```diff
@@ -0,0 +1,281 @@
+                progressHandler: null,
+                buffer: mockBuffer.Object);
+
+            mockBuffer.SetupSequence(r => r.WriteAsync(
+                It.IsAny<byte[]>(),
+                It.IsAny<int>(),
+                It.IsAny<int>(),
+                It.IsAny<CancellationToken>()))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(0))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(0))
+                    .Returns(Task.FromResult(writeSize));
```
What verification is being setup here? `WriteAsync` returns a `Task`, not a `Task<int>`.

seanmcc-msft replied:
Removed in next iteration.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/tests/StorageWriteStreamTests.cs:
```diff
@@ -0,0 +1,281 @@
+                    .Returns(Task.FromResult(0))
+                    .Returns(Task.FromResult(writeSize));
+
+            mockBuffer.SetupSequence(r => r.Position)
+                .Returns(0)
+                .Returns(256)
+                .Returns(512)
+                .Returns(768)
+                .Returns(1024)
+                .Returns(1024)
+                .Returns(0)
+                .Returns(256)
+                .Returns(512)
+                .Returns(1024)
+                .Returns(1024);
```
Does this imply that the `Position` of our write stream reflects where in the buffer we write the next piece of data? If I've written 2048 bytes to a stream then I expect `Position` to return 2048, not whatever index of the internal buffer we happen to be at.

seanmcc-msft committed:
PR comments

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1390,6 +1390,131 @@ public new AppendBlobClient WithVersion(string versionId)
+
+                position = blobPropertiesResponse.Value.ContentLength;
+
+                AppendBlobRequestConditions conditions = new AppendBlobRequestConditions
+                {
+                    IfMatch = blobPropertiesResponse.Value.ETag,
+                    LeaseId = options?.Conditions?.LeaseId,
+                    IfAppendPositionEqual = options?.Conditions?.IfAppendPositionEqual,
+                    IfMaxSizeLessThanOrEqual = options?.Conditions?.IfMaxSizeLessThanOrEqual
+                };
```
Why are we honoring some of the conditions passed in like MaxSize but not others?  What if a tag condition changes halfway through writing?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            await _blockBlobClient.CommitBlockListInternal(
```
You need to update the conditions.  Committing will change the etag so anyone who flushes will then fail their next write.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
```
Nit - maybe just say `greater than zero`.

seanmcc-msft replied:
Fixed in next iteration.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+
+            if (bufferSize > Constants.Blob.Block.MaxStageBytes)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), $"Must <= {Constants.Blob.Block.MaxStageBytes}");
```
Nit - `Must be less than or equal to {...}`

seanmcc-msft replied:
Fixed in next iteration.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+
+                string blockId = StorageExtensions.GenerateBlockId(_position + _buffer.Length);
+
+                if (async)
+                {
+                    await _blockBlobClient.StageBlockAsync(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _blockBlobClient.StageBlock(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
```
Nit - maybe change the visibility of `StageBlockInternal` to keep these calls simpler?

seanmcc-msft replied:
Fixed in next iteration.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs:
```diff
@@ -0,0 +1,99 @@
+                async: async,
+                cancellationToken: cancellationToken)
+                .ConfigureAwait(false);
+        }
```
Do we need to clear _blocksIds here?

seanmcc-msft replied:
No, if we clear block IDs, will will remove all previously created data.

seanmcc-msft replied:
Ted, sorry I accidentally edited your comment when I was trying to respond.

seanmcc-msft replied:
`BlockBlobClient.CommitBlockList()` always commits blocks in the block list as "Latest".  When I was experimenting with this a few days ago, it appears that blocks not included in the block list are removed.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobWriteStream.cs:
```diff
@@ -0,0 +1,74 @@
+                    .ConfigureAwait(false);
+
+                _conditions.IfMatch = response.Value.ETag;
+                _conditions.IfAppendPositionEqual = null;
+                _conditions.IfMaxSizeLessThanOrEqual = null;
```
Why should MaxSize be cleared?

seanmcc-msft replied:
As we are appending, we could eventually hit MaxSize.  The question is if a user sets MaxSize, do they intend for it to be enforced just when we begin writing, or continue to enforce it?  "Let's do an open write if size < MaxSize" vs. "Open write until you hit MaxSize".

seanmcc-msft replied:
Removed in next iteration.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobWriteStream.cs:
```diff
@@ -0,0 +1,74 @@
+                _conditions.IfMaxSizeLessThanOrEqual = null;
+            }
+
+            _buffer.Clear();
```
Any reason to clear it if the length is zero?

seanmcc-msft replied:
No, fixed in next iteration.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobWriteStream.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
```
Make the error messages match what you use in the base class

seanmcc-msft replied:
Changed to `throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be greater than or equal to 1");`.  Note that we are validating the buffer size here, not the `offset` or `count`.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobWriteStream.cs:
```diff
@@ -0,0 +1,74 @@
+
+            if (bufferSize > Constants.Blob.Append.MaxAppendBlockBytes)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must <= 4 MB");
```
Use `Constants.Blob.Append.MaxAppendBlockBytes` in the message so this doesn't get out of date.

seanmcc-msft replied:
Fixed in next iteration.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -2887,6 +2887,142 @@ public new PageBlobClient WithVersion(string versionId)
+            {
+                scope.Start();
+
+                Response<BlobProperties> response = await GetPropertiesInternal(
```
What if it doesn't exist?

seanmcc-msft replied:
We will throw a 404, by design.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,144 @@
+            }
+            else
+            {
+                // We need a multiple of 512 to flush.
+                if (_buffer.Length % Constants.Blob.Page.PageSize != 0)
+                {
+                    int bytesToWrite = (int)(Constants.Blob.Page.PageSize - _buffer.Length % Constants.Blob.Page.PageSize);
+                    await WriteToBufferInternal(buffer, offset, bytesToWrite, async, cancellationToken).ConfigureAwait(false);
+                    remaining -= bytesToWrite;
+                    offset += bytesToWrite;
+                }
```
This is the only part that's different than the base class?  Could we split out another virtual method that we can stick this into so we don't have to duplicate the base class method here?

seanmcc-msft replied:
How would this work?  Pass in the method as a lambda from `PageBlobClient.OpenWrite()`?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,144 @@
+            }
+        }
+
+        protected override async Task FlushInternal(bool async, CancellationToken cancellationToken)
+            => await AppendInternal(async, cancellationToken).ConfigureAwait(false);
```
Should/could this be the default implementation and we only override when we need to replace?

seanmcc-msft replied:
It was originally, but @kasobol-msft asked me to change it.

kasobol-msft replied:
I didn't like idea of underlying abstract type assuming it's "append blob". That would make it half-abstract. In case we ever add new blob type I'd rather have compiler remind the authors that they have to make conscious decision what Flush means for given blob type rather than surprising them.
It costs 1-2 extra lines but I think it's better abstraction.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,144 @@
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
```
Nit - make the message format match the base class

seanmcc-msft replied:
Changed to `throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be greater than or equal to 1");`.  Note that we are checking the buffer size here, not `offset` or `count`.

tg-msft reviewed (commented):
Does the Java implementation share the exact semantics around which conditions are used, updated, and cleared at various points in the life cycle?  What about Track 1?

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/StorageWriteStream.cs:
```diff
@@ -0,0 +1,204 @@
+
+        public override bool CanRead => false;
+
+        public override bool CanSeek => false;
```
We plan to add seekability in the future for Page Blobs and SMB files.  It is not possible to add seekability for Append Blobs, Block Blobs, and Data Lake Files.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -211,6 +211,13 @@ public partial class AppendBlobCreateOptions
     }
+    public partial class AppendBlobOpenWriteOptions
+    {
+        public AppendBlobOpenWriteOptions() { }
```
We decided to remove the overwrite option, and never overwrite except for Block Blobs.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -1257,6 +1281,8 @@ public partial class BlockBlobClient : Azure.Storage.Blobs.Specialized.BlobBaseC
         public virtual Azure.Response<Azure.Storage.Blobs.Models.BlockList> GetBlockList(Azure.Storage.Blobs.Models.BlockListTypes blockListTypes = Azure.Storage.Blobs.Models.BlockListTypes.All, string snapshot = null, Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null ...
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.BlockList>> GetBlockListAsync(Azure.Storage.Blobs.Models.BlockListTypes blockListTypes = Azure.Storage.Blobs.Models.BlockListTypes.All, string snapshot = null, Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threadin ...
+        public virtual System.IO.Stream OpenWrite(Azure.Storage.Blobs.Models.BlockBlobOpenWriteOptions options = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual System.Threading.Tasks.Task<System.IO.Stream> OpenWriteAsync(Azure.Storage.Blobs.Models.BlockBlobOpenWriteOptions options = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
We considered it, but chose not to.  This is an advanced feature, and I think BlobClient is intended for less sophisticated customers.  @jaschrep-msft and I previously discussed this.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -2002,6 +2003,136 @@ protected sealed override BlobBaseClient WithSnapshotCore(string snapshot)
+
+                long position = 0;
+
+                Response<BlobContentInfo> response = await UploadInternal(
+                    content: new MemoryStream(Array.Empty<byte>()),
+                    blobHttpHeaders: default,
+                    metadata: default,
+                    tags: default,
+                    conditions: options?.Conditions,
```
It is not be possible to support seekability for Block Blobs.  In addition, it is not possible for us to Open Write to a Block Blob without overwriting, because Block Blobs created with Put Blob do not contain blocks, so it is not possible for us to keep the data.

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/StorageWriteStream.cs:
```diff
@@ -0,0 +1,204 @@
+            }
+        }
+
+        protected abstract Task FlushInternal(bool async, CancellationToken cancellationToken);
```
We already have an abstract `AppendInternal` method.  This is by design.  Block Blobs and Data Lake Files are updated with an Append/Flush model.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1390,6 +1390,131 @@ public new AppendBlobClient WithVersion(string versionId)
+
+                long position = 0;
+
+                Response<BlobProperties> blobPropertiesResponse = await GetPropertiesInternal(
```
We chose to never overwrite for Append Blob, Page Blob, SMB File, and Data Lake File, and always overwite for Block Blob.

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/Constants.cs:
```diff
@@ -180,6 +180,11 @@ internal static class Block
 
+            internal static class Page
+            {
+                public const int PageSize = 512;
```
Fixed in next iteration.

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:
```diff
@@ -225,7 +230,7 @@ private static async Task<int> ReadLoopInternal(Stream stream, byte[] buffer, in
 
         public override bool CanSeek => true;
 
-        public override bool CanWrite => false;
+        public override bool CanWrite => true;
```
I don't think so. @jaschrep-msft

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/StorageWriteStream.cs:
```diff
@@ -0,0 +1,204 @@
+                throw new ArgumentOutOfRangeException($"{nameof(offset)} cannot be less than 0.");
+            }
+
+            if (offset > buffer.Length)
```
Changed to `if (offset >= buffer.Length)` and updated message in next iteration.

seanmcc-msft committed:
PR comments

seanmcc-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/StorageWriteStream.cs:
```diff
@@ -0,0 +1,204 @@
+                // We need to loop, because remaining bytes might be greater than _buffer size.
+                while (remaining > 0)
+                {
+                    await WriteToBufferInternal(
+                        buffer,
+                        offset,
+                        (int)Math.Min(remaining, _bufferSize),
+                        async,
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Renaming bytes won't fit in buffer.
+                    if (remaining > _bufferSize)
+                    {
+                        await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+                        remaining -= (int)_bufferSize;
+                        offset += (int)_bufferSize;
+                    }
+
+                    // Remaining bytes will fit in buffer.
+                    else
+                    {
+                        remaining = 0;
+                    }
```
Fixed in next iteration.

seanmcc-msft committed:
PR comments

seanmcc-msft committed:
Hopefully fixed CI

seanmcc-msft committed:
Updated Append Blob Open Write

seanmcc-msft committed:
Updated Page Blob Open Write

seanmcc-msft committed:
Updated Block Blob Open Write

seanmcc-msft committed:
Updated Data Lake File Client Open Write

seanmcc-msft committed:
Updated Share File Client Open Write

seanmcc-msft committed:
Ran Export-API.ps1

seanmcc-msft committed:
Recorded tests

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1390,6 +1390,180 @@ public new AppendBlobClient WithVersion(string versionId)
+                {
+                    try
+                    {
+                        Response<BlobProperties> propertiesResponse = await GetPropertiesInternal(
```
`CreateIfNotExists()` wipes out the user's `AccessConditions`, so I can't use it here.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -2887,6 +2887,224 @@ public new PageBlobClient WithVersion(string versionId)
+                {
+                    try
+                    {
+                        Response<BlobProperties> propertiesResponse = await GetPropertiesInternal(
```
`CreateIfNotExists()` wipes out the user's `AccessConditions`, so I can't use it here.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:
```diff
@@ -3864,6 +3877,193 @@ Response<Blobs.Models.BlobDownloadInfo> response
+                }
+                else
+                {
+                    try
```
`CreateIfNotExists()` wipes out the user's `AccessConditions`, so I can't use it here.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4750,5 +4750,209 @@ private void SetNameFieldsIfNull()
+        {
+            DiagnosticScope scope = ClientDiagnostics.CreateScope($"{nameof(ShareFileClient)}.{nameof(OpenWrite)}");
+
+            try
```
`CreateIfNotExists()` wipes out the user's `AccessConditions`, so I can't use it here.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -2887,6 +2887,224 @@ public new PageBlobClient WithVersion(string versionId)
+#pragma warning restore AZC0015 // Unexpected client method return type.
+            bool overwrite,
+            long position,
+            long? size = default,
```
I chose to add an optional size parameter, and mentioned in the xml comments it is required if overwrite is set to true, or the Page Blob is being created for the first time.

seanmcc-msft replied:
Moved `Size` to `PageBlobOpenWriteOptions`.  I'd like to keep `position` a required parameter to prevent users from overwriting the data at the beginning of a Page Blob by default`.

seanmcc-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:
```diff
@@ -4750,5 +4750,209 @@ private void SetNameFieldsIfNull()
+        private async Task<Stream> OpenWriteInternal(
+            bool overwrite,
+            long position,
+            long? maxSize,
```
I chose to add an optional maxSize parameter, and mentioned in the xml comments it is required if overwrite is set to true, or the File is being created for the first time.

seanmcc-msft replied:
Fixed in next iteration.

seanmcc-msft committed:
CI

tg-msft commented on sdk/storage/Azure.Storage.Files.Shares/src/ShareFileWriteStream.cs:
```diff
@@ -0,0 +1,83 @@
+
+                HttpRange httpRange = new HttpRange(_writeIndex, _buffer.Length);
+
+                if (async)
+                {
+                    await _fileClient.UploadRangeAsync(
+                        range: httpRange,
+                        content: _buffer,
+                        progressHandler: _progressHandler,
+                        conditions: _conditions,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _fileClient.UploadRange(
+                        range: httpRange,
+                        content: _buffer,
+                        progressHandler: _progressHandler,
+                        conditions: _conditions,
+                        cancellationToken: cancellationToken);
+                }
```
Don't these need to update the etag?

seanmcc-msft replied:
The Files Service doesn't support ETags.

tg-msft reviewed (commented):
This is awesome progress - I think there are just a couple of other spots we need to update the etag.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -2002,6 +2003,156 @@ protected sealed override BlobBaseClient WithSnapshotCore(string snapshot)
+
+                if (!overwrite)
+                {
+                    throw new ArgumentException($"{nameof(BlockBlobClient)}.{nameof(BlockBlobClient.OpenWrite)} only supports overwrite.");
```
Changed to "only supports overwriting".

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs:
```diff
@@ -0,0 +1,144 @@
+            }
+            else
+            {
+                // We need a multiple of 512 to flush.
+                if (_buffer.Length % Constants.Blob.Page.PageSizeBytes != 0)
+                {
+                    int bytesToWrite = (int)(Constants.Blob.Page.PageSizeBytes - _buffer.Length % Constants.Blob.Page.PageSizeBytes);
+                    await WriteToBufferInternal(buffer, offset, bytesToWrite, async, cancellationToken).ConfigureAwait(false);
+                    remaining -= bytesToWrite;
+                    offset += bytesToWrite;
+                }
+
+                // Flush the buffer.
+                await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+                while (remaining > 0)
+                {
+                    await WriteToBufferInternal(
+                        buffer,
+                        offset,
+                        (int)Math.Min(remaining, _bufferSize),
+                        async,
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Remaining bytes won't fit in buffer.
+                    if (remaining > _bufferSize)
+                    {
+                        await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+                        remaining -= (int)_bufferSize;
+                        offset += (int)_bufferSize;
+                    }
+
+                    // Remaining bytes will fit in buffer.
+                    else
+                    {
+                        remaining = 0;
+                    }
```
No, because the number of bytes we are writing to the buffer is `(int)Math.Min(remaining, _bufferSize),`.  In the case `remaining > _bufferSize`, we just wrote `_bufferSize` to the buffer, which is an increment of 512.

seanmcc-msft commented:
> It looks like we're still not pulling the updated etag after committing the block list? We should add a test with multiple "append, then flush, then append, then append some more, then flush" actions to verify that this works correctly for all blobs.

Fixed in next iteration.

seanmcc-msft commented:
> https://docs.microsoft.com/en-us/rest/api/storageservices/put-block-list says:
>> You can call Put Block List to update a blob by uploading only those blocks that have changed, then committing the new and existing blocks together. You can do this by specifying whether to commit a block from the committed block list or from the uncommitted block list, or to commit the most recently uploaded version of the block, whichever list it may belong to.

`BlockBlobClient.CommitBlockList()` always commits blocks in the block list as "Latest".  When I was experimenting with this a few days ago, it appears that blocks not included in the block list are removed.

seanmcc-msft commented:
> It looks like we're still not pulling the updated etag after committing the block list? We should add a test with multiple "append, then flush, then append, then append some more, then flush" actions to verify that this works correctly for all blobs.

Fixed `BlockBlobWriteStream` to update `IfMatch`, and added a test for this case for all object types.

seanmcc-msft committed:
PR comments

tg-msft reviewed (approved):
Looks great.  Thanks Sean!

yahorsi commented:
Guys stupid question, is this functionality released? I can't find the OpenWrite method on the BlobClient.
Please point me to the right direction

seanmcc-msft commented:
> Guys stupid question, is this functionality released? I can't find the OpenWrite method on the BlobClient.
> Please point me to the right direction

It's not present on the `BlobClient`, it is on `Azure.Storage.Blobs.Specialized.BlockBlobClient`.
